File: ios/DesioraAI/ViewModels/AuthViewModel.swift
import Foundation
import SwiftUI
import AuthenticationServices
import GoogleSignIn

@MainActor
class AuthViewModel: ObservableObject {
    @Published var user: UserResponse?
    @Published var isAuthenticated = false
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let authService = AuthService.shared
    private let tokenStorage = TokenStorage.shared
    
    func checkAuthentication() {
        if tokenStorage.hasTokens() {
            Task {
                await loadUser()
            }
        }
    }
    
    func login(email: String, password: String) async {
        isLoading = true
        errorMessage = nil
        
        do {
            let user = try await authService.login(email: email, password: password)
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func register(email: String, password: String, username: String?) async {
        isLoading = true
        errorMessage = nil
        
        do {
            let user = try await authService.register(email: email, password: password, username: username)
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func signInWithApple(authorization: ASAuthorization) async {
        guard let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential,
              let authorizationCode = appleIDCredential.authorizationCode,
              let codeString = String(data: authorizationCode, encoding: .utf8) else {
            errorMessage = "Failed to get Apple authorization code"
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        do {
            let user = try await authService.signInWithApple(authorizationCode: codeString)
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func signInWithGoogle() async {
        guard let windowScene = await UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let rootViewController = windowScene.windows.first?.rootViewController else {
            errorMessage = "Unable to get root view controller"
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        do {
            let result = try await GIDSignIn.sharedInstance.signIn(withPresenting: rootViewController)
            guard let idToken = result.user.idToken?.tokenString else {
                throw NSError(domain: "GoogleSignIn", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to get ID token"])
            }
            
            let user = try await authService.signInWithGoogle(idToken: idToken)
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func loadUser() async {
        do {
            let user = try await authService.getCurrentUser()
            self.user = user
            self.isAuthenticated = true
        } catch {
            self.user = nil
            self.isAuthenticated = false
        }
    }
    
    func logout() {
        do {
            try authService.logout()
            self.user = nil
            self.isAuthenticated = false
        } catch {
            self.errorMessage = error.localizedDescription
        }
    }
}

File: android/app/src/main/java/com/desiora/ai/ui/viewmodel/AuthViewModel.kt
package com.desiora.ai.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.desiora.ai.data.model.UserResponse
import com.desiora.ai.data.repository.AuthRepository
import com.desiora.ai.data.storage.TokenStorage
import kotlinx.coroutines.launch

class AuthViewModel(private val tokenStorage: TokenStorage) : ViewModel() {
    
    private val authRepository = AuthRepository(tokenStorage)
    
    private val _user = MutableLiveData<UserResponse?>()
    val user: LiveData<UserResponse?> = _user
    
    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading
    
    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> = _error
    
    fun login(email: String, password: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            authRepository.login(email, password)
                .onSuccess { user ->
                    _user.value = user
                    _isLoading.value = false
                }
                .onFailure { exception ->
                    _error.value = exception.message
                    _isLoading.value = false
                }
        }
    }
    
    fun register(email: String, password: String, username: String?) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            authRepository.register(email, password, username)
                .onSuccess { user ->
                    _user.value = user
                    _isLoading.value = false
                }
                .onFailure { exception ->
                    _error.value = exception.message
                    _isLoading.value = false
                }
        }
    }
    
    fun googleOAuth(code: String, redirectUri: String?) {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            authRepository.googleOAuth(code, redirectUri)
                .onSuccess { user ->
                    _user.value = user
                    _isLoading.value = false
                }
                .onFailure { exception ->
                    _error.value = exception.message
                    _isLoading.value = false
                }
        }
    }
    
    fun getCurrentUser() {
        viewModelScope.launch {
            _isLoading.value = true
            authRepository.getCurrentUser()
                .onSuccess { user ->
                    _user.value = user
                    _isLoading.value = false
                }
                .onFailure {
                    _user.value = null
                    _isLoading.value = false
                }
        }
    }
    
    fun logout() {
        authRepository.logout()
        _user.value = null
    }
    
    fun isLoggedIn(): Boolean {
        return authRepository.isLoggedIn()
    }
    
    fun clearError() {
        _error.value = null
    }
}


